django/test/selenium.py

```diff
@@ -134,3 +137,82 @@ def disable_implicit_wait(self):
             yield
         finally:
             self.selenium.implicitly_wait(self.implicit_wait)
+
+
+def screenshot_cases(method_names):
+    if isinstance(method_names, str):
+        method_names = method_names.split(",")
+
+    def wrapper(func):
+        func._screenshot_cases = method_names
+        return func
+
+    return wrapper
+
+
+@tag("screenshot")
+class ScreenshotTestCase(SeleniumTestCase):
+    screenshots = False
+
+    @classmethod
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        if not cls.screenshots:
+            return
+
+        for name, func in list(cls.__dict__.items()):
+            if not hasattr(func, "_screenshot_cases"):
+                continue
+
+            delattr(cls, name)
+            _screenshot_cases = getattr(func, "_screenshot_cases")
+            for screenshot_case in _screenshot_cases:
+
+                @wraps(func)
+                def test(self, *args, _func=func, _case=screenshot_case, **kwargs):
+                    with getattr(self, _case)():
+                        return _func(self, *args, **kwargs)
+
+                test.__name__ = f"{name}_{screenshot_case}"
+                test.__qualname__ = f"{test.__qualname__}_{screenshot_case}"
+
+                setattr(cls, test.__name__, test)
+
+    @contextmanager
+    def desktop(self):
+        window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(1280, 720)
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+
+    @contextmanager
+    def mobile(self):
+        window_size = self.selenium.get_window_size()
+        self.selenium.set_window_size(360, 800)
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+
+    @contextmanager
+    def rtl(self):
+        window_size = self.selenium.get_window_size()
+        get_language_bidi_mock = mock.patch(
+            "django.utils.translation.get_language_bidi", return_value=True
+        )
+        self.selenium.set_window_size(1280, 720)
+        get_language_bidi_mock.start()
+        try:
+            yield
+        finally:
+            self.selenium.set_window_size(window_size["width"], window_size["height"])
+            get_language_bidi_mock.stop()
```

Comment:
I wonder if these 3 context managers could be moved to `SeleniumTestCase` and used as decorators directly on test methods (we have some tests that use `set_window_size()`), and reused here. This could standardize tests.

Also, we should use the same boundary widths as defined in admin CSS files, i.e.
- <= 767 (mobile)
- <= 1024 (small screens)
- standard

so maybe:

- `(767, 800)` -> `@mobile`
- `(1024, 768)` -> `@small_screen`
- `(1280, 720)` -> `@desktop`

We can do this in a separate PR.

Reply:
So I would start with adding four context processors that can be used as test decorators, `@mobile`, `@small_screen`, `@desktop`, and `@rtl`.

Reply:
Interesting suggestion. Originally I assumed these are only used/wanted when doing screenshots, if you agree with that then do you think the naming should be more exclicit around that it's related to taking a screenshot if this was updated? (like `@screenshot_mobile`)
If you think these are useful just in general, then I guess when doing the standard selenium tests we would want to re-run the test for each case regardless of whether it is taking a screenshot 🤔

Reply:
> If you think these are useful just in general, then I guess when doing the standard selenium tests we would want to re-run the test for each case regardless of whether it is taking a screenshot 🤔

Yes, that's why I left this suggestion :smile: I imagine something like:
```python
class ChangeWindowSize:
    def __init__(self, width, height, selenium):
        self.new_size = (width, height)
        self.old_size = tuple(self.selenium.get_window_size().values())

    def __enter__(self):
        self.old_size = self.selenium.get_window_size()
        self.selenium.set_window_size(*self.new_size)
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.selenium.set_window_size(*self.old_size)

class SeleniumTestCase(...)

    @contextmanager
    def desktop(self):
        with ChangeWindowSize(1280, 720, self.selenium):
            yield

    @contextmanager
    def small_screen(self):
        with ChangeWindowSize(1024, 768, self.selenium):
            yield

    ...
```
and in tests, e.g.
```diff
diff --git a/tests/admin_widgets/tests.py b/tests/admin_widgets/tests.py
index 80bb6233ab..5ef69ce049 100644
--- a/tests/admin_widgets/tests.py
+++ b/tests/admin_widgets/tests.py
@@ -1076,6 +1076,7 @@ class DateTimePickerSeleniumTests(AdminWidgetSeleniumTestCase):
 
         self.assertEqual(len(selected), 0)
 
+    @self.small_screen
     def test_calendar_show_date_from_input(self):
         """
         The calendar shows the date from the input field for every locale
@@ -1083,7 +1084,6 @@ class DateTimePickerSeleniumTests(AdminWidgetSeleniumTestCase):
         """
         from selenium.webdriver.common.by import By
 
-        self.selenium.set_window_size(1024, 768)
         self.admin_login(username="super", password="secret", login_url="/")
 
         # Enter test data
```

The same context processors would be reused later in `_screenshot_cases`.

Reply:
Ah thank you, I see what you're going for 👍